// nuxtseo-shared/dist/const.mjs
const NuxtSEO = {
  slug: "nuxt-seo",
  label: "Nuxt SEO",
  icon: "i-carbon-3rd-party-connected",
  description: "The all-in-one module that brings it all together.",
  repo: "harlan-zw/nuxt-seo",
  npm: "@nuxtjs/seo",
  playgrounds: {
    basic: "https://stackblitz.com/github/harlan-zw/nuxt-seo/tree/main/examples/basic",
    i18n: "https://stackblitz.com/github/harlan-zw/nuxt-seo/tree/main/examples/i18n"
  }
};
const SiteConfigModule = {
  slug: "site-config",
  label: "Site Config",
  icon: "i-carbon-settings-check",
  description: "Powerful build and runtime shared site configuration for Nuxt modules.",
  repo: "harlan-zw/nuxt-site-config",
  npm: "nuxt-site-config",
  playgrounds: {
    "basic": "https://stackblitz.com/github/harlan-zw/nuxt-site-config/tree/main/examples/basic",
    "env-driven": "https://stackblitz.com/github/harlan-zw/nuxt-site-config/tree/main/examples/env-driven",
    "multi-site": "https://stackblitz.com/github/harlan-zw/nuxt-site-config/tree/main/examples/multi-site"
  }
};
const RobotsModule = {
  slug: "robots",
  label: "Robots",
  icon: "i-carbon-bot",
  description: "Tame the robots crawling and indexing your site with ease.",
  repo: "nuxt-modules/robots",
  npm: "@nuxtjs/robots",
  playgrounds: {
    "basic": "https://stackblitz.com/github/nuxt-modules/robots/tree/main/examples/basic",
    "i18n": "https://stackblitz.com/github/nuxt-modules/robots/tree/main/examples/i18n",
    "custom-rules": "https://stackblitz.com/github/nuxt-modules/robots/tree/main/examples/custom-rules"
  }
};
const SitemapModule = {
  slug: "sitemap",
  label: "Sitemap",
  icon: "i-carbon-load-balancer-application",
  description: "Powerfully flexible XML Sitemaps that integrate seamlessly.",
  repo: "nuxt-modules/sitemap",
  npm: "@nuxtjs/sitemap",
  playgrounds: {
    "basic": "https://stackblitz.com/github/nuxt-modules/sitemap/tree/main/examples/basic",
    "i18n": "https://stackblitz.com/github/nuxt-modules/sitemap/tree/main/examples/i18n",
    "dynamic-urls": "https://stackblitz.com/github/nuxt-modules/sitemap/tree/main/examples/dynamic-urls"
  }
};
const OgImageModule = {
  slug: "og-image",
  label: "OG Image",
  icon: "i-carbon-image-search",
  description: "Generate OG Images with Vue templates in Nuxt.",
  repo: "nuxt-modules/og-image",
  npm: "nuxt-og-image",
  playgrounds: {
    "basic-satori": "https://stackblitz.com/github/nuxt-modules/og-image/tree/main/examples/basic-satori",
    "basic-takumi": "https://stackblitz.com/github/nuxt-modules/og-image/tree/main/examples/basic-takumi",
    "content": "https://stackblitz.com/github/nuxt-modules/og-image/tree/main/examples/content",
    "i18n": "https://stackblitz.com/github/nuxt-modules/og-image/tree/main/examples/i18n"
  }
};
const LinkCheckerModule = {
  slug: "link-checker",
  label: "Link Checker",
  icon: "i-carbon-cloud-satellite-link",
  description: "Find and magically fix links that may be negatively effecting your SEO.",
  repo: "harlan-zw/nuxt-link-checker",
  npm: "nuxt-link-checker",
  playgrounds: {
    "basic": "https://stackblitz.com/github/harlan-zw/nuxt-link-checker/tree/main/examples/basic",
    "broken-links": "https://stackblitz.com/github/harlan-zw/nuxt-link-checker/tree/main/examples/broken-links",
    "skip-inspection": "https://stackblitz.com/github/harlan-zw/nuxt-link-checker/tree/main/examples/skip-inspection"
  }
};
const SeoUtilsModule = {
  slug: "seo-utils",
  label: "SEO Utils",
  icon: "i-carbon-tools",
  description: "SEO utilities to improve your Nuxt sites discoverability and shareability.",
  repo: "harlan-zw/nuxt-seo-utils",
  npm: "nuxt-seo-utils",
  playgrounds: {
    "basic": "https://stackblitz.com/github/harlan-zw/nuxt-seo-utils/tree/main/examples/basic",
    "breadcrumbs": "https://stackblitz.com/github/harlan-zw/nuxt-seo-utils/tree/main/examples/breadcrumbs",
    "meta-tags": "https://stackblitz.com/github/harlan-zw/nuxt-seo-utils/tree/main/examples/meta-tags"
  }
};
const SchemaOrgModule = {
  slug: "schema-org",
  label: "Schema.org",
  icon: "i-carbon-chart-relationship",
  description: "The quickest and easiest way to build Schema.org graphs.",
  repo: "harlan-zw/nuxt-schema-org",
  npm: "nuxt-schema-org",
  playgrounds: {
    "basic": "https://stackblitz.com/github/harlan-zw/nuxt-schema-org/tree/main/examples/basic",
    "blog": "https://stackblitz.com/github/harlan-zw/nuxt-schema-org/tree/main/examples/blog",
    "e-commerce": "https://stackblitz.com/github/harlan-zw/nuxt-schema-org/tree/main/examples/e-commerce"
  }
};
const SkewProtectionModule = {
  slug: "skew-protection",
  npm: "nuxt-skew-protection",
  repo: "nuxt-seo-pro/nuxt-skew-protection",
  description: "Solve Nuxt version skews with persistent assets and instant updates.",
  label: "Skew Protection",
  icon: "i-carbon-version"
};
const AiReadyModule = {
  slug: "ai-ready",
  npm: "nuxt-ai-ready",
  repo: "nuxt-seo-pro/nuxt-ai-ready",
  description: "Best practice AI & LLM discoverability for Nuxt sites.",
  label: "AI Ready",
  icon: "i-carbon-ai-label"
};
const modules = [
  NuxtSEO,
  RobotsModule,
  SitemapModule,
  OgImageModule,
  SchemaOrgModule,
  LinkCheckerModule,
  SeoUtilsModule,
  SiteConfigModule,
  SkewProtectionModule,
  AiReadyModule
];
const bundledModules = [
  RobotsModule,
  SitemapModule,
  OgImageModule,
  SchemaOrgModule,
  LinkCheckerModule,
  SeoUtilsModule
];
const standaloneModules = [
  SkewProtectionModule,
  AiReadyModule
];
const normalModules = bundledModules;
const proModules = standaloneModules;

export { AiReadyModule, LinkCheckerModule, NuxtSEO, OgImageModule, RobotsModule, SchemaOrgModule, SeoUtilsModule, SiteConfigModule, SitemapModule, SkewProtectionModule, bundledModules, modules, normalModules, proModules, standaloneModules };
