// nuxtseo-shared/dist/kit.mjs
import { pathToFileURL } from 'node:url';
import { useNuxt, createResolver, addTemplate, loadNuxtModuleInstance, tryUseNuxt, hasNuxtModule, hasNuxtModuleCompatibility } from '@nuxt/kit';
import { relative, dirname } from 'pathe';
import { resolvePackageJSON, readPackageJSON } from 'pkg-types';
import { provider, env } from 'std-env';

function normalizePackageUrl(rootDir) {
  const url = rootDir.startsWith("file:") ? rootDir : pathToFileURL(rootDir.endsWith("/") ? rootDir : `${rootDir}/`).href;
  return url.endsWith("/") ? url : `${url}/`;
}
const NUXT_SEO_MODULES = /* @__PURE__ */ new Set([
  "@nuxtjs/robots",
  "@nuxtjs/sitemap",
  "nuxt-og-image",
  "nuxt-schema-org",
  "nuxt-seo-utils",
  "nuxt-link-checker",
  "nuxt-site-config",
  "nuxt-skew-protection",
  "nuxt-ai-ready"
]);
function detectNuxtSeoModules(nuxt = useNuxt()) {
  return nuxt.options._installedModules.filter((m) => m.meta?.name && NUXT_SEO_MODULES.has(m.meta.name)).map((m) => ({
    name: m.meta.name,
    version: m.meta.version,
    entryPath: m.entryPath
  }));
}
const autodetectableProviders = {
  azure_static: "azure",
  cloudflare_pages: "cloudflare-pages",
  netlify: "netlify",
  stormkit: "stormkit",
  vercel: "vercel",
  cleavr: "cleavr",
  stackblitz: "stackblitz"
};
const autodetectableStaticProviders = {
  netlify: "netlify-static",
  vercel: "vercel-static"
};
function detectTarget(options = {}) {
  return options?.static ? autodetectableStaticProviders[provider] : autodetectableProviders[provider];
}
function resolveNitroPreset(nitroConfig) {
  nitroConfig = nitroConfig || tryUseNuxt()?.options?.nitro;
  if (provider === "stackblitz" || provider === "codesandbox")
    return provider;
  let preset;
  if (nitroConfig && nitroConfig?.preset)
    preset = nitroConfig.preset;
  if (!preset)
    preset = env.NITRO_PRESET || env.SERVER_PRESET || detectTarget() || "node-server";
  return preset.replaceAll("_", "-");
}
async function getNuxtModuleOptions(module, nuxt = useNuxt()) {
  const moduleMeta = (typeof module === "string" ? { name: module } : await module.getMeta?.()) || {};
  const { nuxtModule } = await loadNuxtModuleInstance(module, nuxt);
  let moduleEntry;
  for (const m of nuxt.options.modules) {
    if (Array.isArray(m) && m.length >= 2) {
      const _module = m[0];
      const _moduleEntryName = typeof _module === "string" ? _module : (await _module.getMeta?.())?.name || "";
      if (_moduleEntryName === moduleMeta.name)
        moduleEntry = m;
    }
  }
  let inlineOptions = {};
  if (moduleEntry)
    inlineOptions = moduleEntry[1];
  if (nuxtModule.getOptions)
    return nuxtModule.getOptions(inlineOptions, nuxt);
  return inlineOptions;
}
function isNuxtGenerate(nuxt = useNuxt()) {
  const nitroOptions = nuxt.options.nitro;
  return nitroOptions?.static || nuxt.options._generate || [
    "static",
    "github-pages"
  ].includes(resolveNitroPreset(nitroOptions));
}
function extendTypes(module, template) {
  const nuxt = useNuxt();
  const { resolve } = createResolver(import.meta.url);
  addTemplate({
    filename: `module/${module}.d.ts`,
    getContents: async () => {
      const typesPath = relative(resolve(nuxt.options.rootDir, nuxt.options.buildDir, "module"), resolve("runtime/types"));
      const s = await template({ typesPath });
      return `// Generated by ${module}
${s}
export {}
`;
    }
  });
  nuxt.hooks.hook("prepare:types", ({ references }) => {
    references.push({ path: resolve(nuxt.options.buildDir, `module/${module}.d.ts`) });
  });
}
function createPagesPromise(nuxt = useNuxt()) {
  return new Promise((resolve) => {
    nuxt.hooks.hook("modules:done", () => {
      if (typeof nuxt.options.pages === "boolean" && nuxt.options.pages === false || typeof nuxt.options.pages === "object" && !nuxt.options.pages.enabled) {
        return resolve([]);
      }
      nuxt.hook("pages:resolved", (pages) => resolve(pages));
    });
  });
}
function createNitroPromise(nuxt = useNuxt()) {
  return new Promise((resolve) => {
    nuxt.hooks.hook("nitro:init", (nitro) => resolve(nitro));
  });
}
async function resolveNuxtContentVersion() {
  if (!hasNuxtModule("@nuxt/content"))
    return false;
  if (await hasNuxtModuleCompatibility("@nuxt/content", "^3"))
    return { version: 3 };
  if (await hasNuxtModuleCompatibility("@nuxt/content", "^2"))
    return { version: 2 };
  return false;
}
async function resolvePackageMajor(id, rootDir) {
  const url = normalizePackageUrl(rootDir);
  const version = await readPackageJSON(id, { url }).then((pkg) => pkg.version).catch(() => {
    return void 0;
  });
  const major = version ? Number.parseInt(version, 10) : Number.NaN;
  return Number.isFinite(major) ? major : void 0;
}
async function resolveHostUnheadMajor(rootDir) {
  const rootUrl = normalizePackageUrl(rootDir);
  const searchUrls = [];
  for (const id of ["@nuxt/nitro-server", "nuxt"]) {
    const pkgJson = await resolvePackageJSON(id, { url: rootUrl }).catch(() => {
      return void 0;
    });
    if (pkgJson)
      searchUrls.push(`${dirname(pkgJson)}/`);
  }
  searchUrls.push(rootUrl);
  for (const id of ["@unhead/vue", "unhead"]) {
    for (const url of searchUrls) {
      const major = await resolvePackageMajor(id, url);
      if (major === 2)
        return 2;
      if (major !== void 0 && major >= 3)
        return 3;
    }
  }
  return 3;
}

export { createNitroPromise, createPagesPromise, detectNuxtSeoModules, detectTarget, extendTypes, getNuxtModuleOptions, isNuxtGenerate, resolveHostUnheadMajor, resolveNitroPreset, resolveNuxtContentVersion, resolvePackageMajor };
