// @nuxtjs/sitemap/dist/runtime/server/routes/__sitemap__/debug-production.js
import { isSitemapIndex, parseSitemapIndex, parseSitemapXml } from "@nuxtjs/sitemap/utils";
import { defineEventHandler, getQuery } from "h3";
async function fetchXml(url) {
  const response = await fetch(url, {
    headers: { Accept: "application/xml, text/xml" },
    signal: AbortSignal.timeout(15e3)
  });
  if (!response.ok)
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  return response.text();
}
export default defineEventHandler(async (e) => {
  const { url, mode } = getQuery(e);
  if (!url || typeof url !== "string")
    return { url: "", isIndex: false, sitemaps: [], warnings: [], error: "Missing url query parameter" };
  if (mode === "debug") {
    const debugUrl = `${url.replace(/\/$/, "")}/__sitemap__/debug.json`;
    const response = await fetch(debugUrl, {
      headers: { Accept: "application/json" },
      signal: AbortSignal.timeout(1e4)
    }).catch(() => null);
    if (response?.ok) {
      const json = await response.json().catch(() => null);
      if (json?.sitemaps)
        return json;
    }
  }
  const sitemapUrl = url.endsWith("/") ? `${url}sitemap.xml` : url;
  const xml = await fetchXml(sitemapUrl).catch((err) => {
    return err;
  });
  if (xml instanceof Error)
    return { url: sitemapUrl, isIndex: false, sitemaps: [], warnings: [], error: `Failed to fetch sitemap: ${xml.message}` };
  if (isSitemapIndex(xml)) {
    const { entries, warnings } = await parseSitemapIndex(xml);
    const sitemaps = await Promise.all(
      entries.map(async (entry) => {
        const childXml = await fetchXml(entry.loc).catch((err) => err);
        if (childXml instanceof Error) {
          return {
            loc: entry.loc,
            urlCount: 0,
            warnings: [],
            error: childXml.message,
            lastmod: entry.lastmod
          };
        }
        const result2 = await parseSitemapXml(childXml).catch((err) => ({
          urls: [],
          warnings: [{ type: "validation", message: err.message }]
        }));
        return {
          loc: entry.loc,
          urlCount: result2.urls.length,
          warnings: result2.warnings,
          lastmod: entry.lastmod
        };
      })
    );
    return { url: sitemapUrl, isIndex: true, sitemaps, warnings };
  }
  const result = await parseSitemapXml(xml).catch((err) => ({
    urls: [],
    warnings: [{ type: "validation", message: err.message }]
  }));
  return {
    url: sitemapUrl,
    isIndex: false,
    sitemaps: [{
      loc: sitemapUrl,
      urlCount: result.urls.length,
      warnings: result.warnings
    }],
    warnings: []
  };
});
